3-enrich/spark-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich.spark/EnrichJob.scala

```diff
@@ -0,0 +1,173 @@
+    classOf[Array[String]],
+    classOf[Array[Array[Byte]]],
+    classOf[EnrichedEvent],
+    classOf[scala.collection.immutable.Map$EmptyMap$],
+    classOf[scala.collection.immutable.Set$EmptySet$],
+    classOf[scala.collection.mutable.WrappedArray$ofRef],
+    classOf[org.apache.spark.internal.io.FileCommitProtocol$TaskCommitMessage]
+  )
+  override def sparkConfig(): SparkConf = new SparkConf()
+    .setAppName(getClass().getSimpleName())
+    .setIfMissing("spark.master", "local[*]")
+    .set("spark.serializer", classOf[KryoSerializer].getName())
+    .set("spark.kryo.registrationRequired", "true")
+    .registerKryoClasses(classesToRegister)
+
+  override def run(spark: SparkSession, args: Array[String]): Unit = {
+    val job = EnrichJob(spark, args)
+    job.run()
+  }
+
+  def apply(spark: SparkSession, args: Array[String]) = new EnrichJob(spark, args)
+
+  val etlVersion = s"spark-${generated.ProjectSettings.version}"
+
+  /**
+   * Project our Failures into a List of Nel of strings.
+   * @param all A List of Validations each containing either an EnrichedEvent or Failure strings
+   * @return A (possibly empty) List of failures, where each failure is a Nel of strings
+   */
+  def projectBads(all: List[ValidatedEnrichedEvent]): List[NonEmptyList[String]] =
+    all.collect { case Failure(errs) => errs }
+
+  /**
+   * Project our Sucesses into EnrichedEvents.
+   * @param all A List of Validations each containing either an EnrichedEvent or Failure strings
+   * @return A (possibly empty) List of EnrichedEvent
+   */
+  def projectGoods(all: List[ValidatedEnrichedEvent]): List[EnrichedEvent] =
+    all.collect { case Success(e) => e }
+}
+
+/**
+ * The Snowplow Enrich job, written in Spark.
+ * @param spark Spark session used throughout the job
+ * @param args Command line arguments for the enrich job
+ */
+class EnrichJob(@transient val spark: SparkSession, args: Array[String]) extends Serializable {
+  @transient private val sc: SparkContext = spark.sparkContext
+  import spark.implicits._
+
+  // Necessary to be able to deal with LZO-compressed files
+  @transient private val hadoopConfig = sc.hadoopConfiguration
+  hadoopConfig.set("io.compression.codecs", classOf[com.hadoop.compression.lzo.LzopCodec].getName())
+  hadoopConfig.set("io.compression.codec.lzo.class",
+    classOf[com.hadoop.compression.lzo.LzoCodec].getName())
+
+  // Job configuration
+  private val enrichConfig = EnrichJobConfig.loadConfigFrom(args).fold(
+    e => throw FatalEtlError(e.map(_.toString)),
+    identity
+  )
+
+  def run(): Unit = {
+    import EnrichJob._
+    import singleton._
+
+    // Install MaxMind file(s) if we have them
+    enrichConfig.filesToCache.foreach { case (uri, filename) => sc.addFile(uri.toString) }
+
+    val inFormat = enrichConfig.inFormat
+    val igluConfig = enrichConfig.igluConfig
+
+    val input = getInputRDD(inFormat, enrichConfig.inFolder)
+
+    val common = input
+      .map { line =>
```

Comment:
Can we refactor contents of this tuple into separate values like this:

```scala
      .map { line =>
        val registry = RegistrySingleton.get(igluConfig, enrichConfig.enrichments, enrichConfig.local)
        val loader = LoaderSingleton.get(inFormat).asInstanceOf[Loader[Any]].toCollectorPayload(line)
        val event = EtlPipeline.processEvents(registry, etlVersion, enrichConfig.etlTstamp, loader)(ResolverSingleton.get(igluConfig))
        (line, event)
      }
```

I think it would be more readable even to factor it out as separate function. Basically this small lambda is the heart of Snowplow.

Also, I see that `registry` and
`Loader[Any]` are singletons created once per worker, but is it really necessary to instantiate them straight inside lambda?